### PR TITLE
Better livestream support

### PR DESCRIPTION
**Pull request by linkjay**
Various changes around to allow the audio module to work better with livestreams

### CODE REVIEW

**Comment by linkjay** [comment]
There's only 1 commit. The merge there was due to the fact that my fork didn't contain the most up-to-date master branch because...github? Merging with that merge *shouldn't* harm anything because the code is already here but who knows

File: plugins/audio.js [hunk 1]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
...
+					//default to false
+					info.isLivestream = false
+
+					//detect if livestream
+					if(info.length_seconds == "0" && info.streamurl.indexOf("yt_live_broadcast") !== -1) {
+						//set a custom var for livestreams
+						info.isLivestream = true
+						info.length = "LIVE"
+
+						if(!settings.get( 'audio', 'allow_youtube_livestreams', false )) {
```
Comment (DougTy) [inline plugins/audio.js]: - please put brace on new line & pad statement with spaces inside `( )`
- brackets not necessary for single-line statement
- change `allow_youtube_livestreams` to `allow_livestreams`

File: plugins/audio.js [hunk 2]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
 		{
 			queryRemote( msg, url ).then( info =>
 				{
+					//default to false
+					info.isLivestream = false
+
+					//detect if livestream
+					if(info.length_seconds == "0" && info.streamurl.indexOf("yt_live_broadcast") !== -1) {
```
Comment (DougTy) [inline plugins/audio.js]: - please put brace on new line & pad statement with spaces inside `( )`
- use `===` instead of `==`
- single quote instead of double
- space after `//`
- remove check for `yt_live_broadcast` so it works with all live streams

File: plugins/audio.js [hunk 3]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
...
+						//set a custom var for livestreams
+						info.isLivestream = true
+						info.length = "LIVE"
```
Comment (DougTy) [inline plugins/audio.js]: single quote

File: plugins/audio.js [hunk 4]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
 		{
 			queryRemote( msg, url ).then( info =>
 				{
+					//default to false
+					info.isLivestream = false
+
+					//detect if livestream
+					if(info.length_seconds == "0" && info.streamurl.indexOf("yt_live_broadcast") !== -1) {
+						//set a custom var for livestreams
```
Comment (DougTy) [inline plugins/audio.js]: unnecessary comment

File: plugins/audio.js [hunk 5]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
...
+						info.isLivestream = true
+						info.length = "LIVE"
+
+						if(!settings.get( 'audio', 'allow_youtube_livestreams', false )) {
+							//is livestream and not allowed
+							return msg.channel.send( "livestreams are not allowed" )
```
Comment (DougTy) [inline plugins/audio.js]: - single quote
- unnecessary comment

File: plugins/audio.js [hunk 6]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
...
+						}
+					}
+
+					//queue up if we're still here
```
Comment (DougTy) [inline plugins/audio.js]: unnecessary comment

File: plugins/audio.js [hunk 7]
```diff
@@ -1017,8 +1033,15 @@ commands.register( {
 									for ( const i in data.items )
 									{
 										const song = data.items[i]
-										fields.push( { name: `${parseInt(i)+1}. ${song.title} [${song.duration}] (${song.author.name})`, value: song.link } )
-										results.push( song.link )
+										
+										//check if its a livestream by checking if duration is set, and only display this entry if we allow them
+										if(song.duration || (settings.get( 'audio', 'allow_youtube_livestreams', false ) && !song.duration)) {
```
Comment (DougTy) [inline plugins/audio.js]: - please put brace on new line & pad statement with spaces inside `( )`
- space after `//`
- change `allow_youtube_livestreams` to `allow_livestreams`

File: plugins/audio.js [hunk 8]
```diff
@@ -1017,8 +1033,15 @@ commands.register( {
...
+											let length = song.duration
+											if(!song.duration) length = "LIVE"
```
Comment (DougTy) [inline plugins/audio.js]: - please pad statement with spaces inside `( )`
- single quote
- declaration on new line

File: plugins/audio.js [hunk 9]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
 		{
 			queryRemote( msg, url ).then( info =>
 				{
+					//default to false
+					info.isLivestream = false
```
Comment (DougTy) [inline plugins/audio.js]: - unnecessary comment
- shouldn't be necessary to set isLivestream to a default value as `if ( song.isLivestream )` will be false if undefined

File: plugins/audio.js [hunk 10]
```diff
@@ -1211,6 +1234,7 @@ commands.register( {
 			let by_user = get_queuedby_user( song )
 			if ( sess.queue.length > 1 )
 				by_user += `, +${sess.queue.length - 1} in queue`
+
```
Comment (DougTy) [inline plugins/audio.js]: unnecessary change

File: plugins/audio.js [hunk 11]
```diff
@@ -1234,19 +1258,33 @@ commands.register( {
...
 			const fields = []
 			for ( const i in queue )
 			{
 				const song = queue[i]
 				total_len += parseInt( song.length_seconds )
 				const by_user = get_queuedby_user( song )
+
+				let length = song.length
+				if(song.isLivestream) {
+					length = "LIVE"
```
Comment (DougTy) [inline plugins/audio.js]: - please put brace on new line & pad statement with spaces inside `( )`
- single quotes

File: plugins/audio.js [hunk 12]
```diff
@@ -1234,19 +1258,33 @@ commands.register( {
...
+					livestreams++
+				} 
+
 				fields.push( { name: _.fmt( '%s. %s [%s] (%s)', parseInt(i) + 1, song.title, song.length, by_user ), value: song.url } )
 			}
-			
+
```
Comment (DougTy) [inline plugins/audio.js]: unnecessary change

File: plugins/audio.js [hunk 13]
```diff
@@ -1234,19 +1258,33 @@ commands.register( {
...
 			total_len = formatTime( total_len )
 
+			if(livestreams != 0) {
+				livestreams = "+ "+ livestreams +" livestreams"
+			}else {
+				livestreams = ""
+			}
+
```
Comment (DougTy) [inline plugins/audio.js]: - please put brace on new line & pad statement with spaces inside `( )`
- brackets not necessary for single-line statements
- could define `livestreams = ''` above condition & avoid using else
- single quotes
- space out `" + concatenation + "`
- use `!==` instead of `!=`

File: plugins/audio.js [hunk 14]
```diff
@@ -1306,14 +1344,17 @@ commands.register( {
 			if ( !sess.playing ) return
 			
 			if ( args )
-				start_player( sess, _.parsetime( args ) )
+				start_player( sess, _.parsetime( args ) )
```
Comment (DougTy) [inline plugins/audio.js]: unnecessary change

File: plugins/audio.js [hunk 15]
```diff
@@ -1306,14 +1344,17 @@ commands.register( {
...
-				start_player( sess, _.parsetime( args ) )
+				start_player( sess, _.parsetime( args ) )	
 			else
 			{
 				let currentSeek = formatTime( Math.round(sess.time) )
 				if ( !currentSeek.match( ':' ) )
 					currentSeek = '00:' + currentSeek
+
+				let length = sess.queue[0].length
+				if(sess.queue[0].isLivestream) length = "FOREVER"
```
Comment (DougTy) [inline plugins/audio.js]: - please pad statement with spaces inside `( )`
- declaration on newline
- single quotes

File: plugins/audio.js [hunk 16]
```diff
@@ -814,6 +814,22 @@ function playURL( url, msg )
 		{
 			queryRemote( msg, url ).then( info =>
 				{
```
Comment (DougTy) [inline plugins/audio.js]: This should all be inside queryRemote, not playURL as playURL isn't used for everything (and queryRemote is)
As it is this will only work for certain commands

File: plugins/audio.js [hunk 17]
```diff
@@ -1234,19 +1258,33 @@ commands.register( {
 				return msg.channel.send( '```\nempty\n```' )
```
Comment (DougTy) [inline plugins/audio.js]: Similar embeds (ie. for playlists) unchanged -- perhaps I should refactor them all to use common code before adding livestream specific code like this. I fear feature creep with added livestream code to every command.

File: plugins/audio.js [hunk 18]
```diff
@@ -1306,14 +1344,17 @@ commands.register( {
 			if ( !sess.playing ) return
 			
 			if ( args )
-				start_player( sess, _.parsetime( args ) )
+				start_player( sess, _.parsetime( args ) )	
 			else
 			{
 				let currentSeek = formatTime( Math.round(sess.time) )
```
Comment (DougTy) [inline plugins/audio.js]: This is a much simpler, easier change but like !queue I fear feature creep trying to add livestream specific code into every command. It might just be better to leave everything as is, else we may have some commands that handle livestreams and some that don't. I'm still undecided

**Comment by DougTy** [comment]
Urgh don't view the code review from here, click on the file, I fucked it up

**Comment by linkjay** [comment]
remaking this pr due to this one being a total mess

**Comment by DougTy** [comment]
> remaking this pr due to this one being a total mess

Sorry, my fault D:
